moveit_setup_assistant/src/widgets/controllers_widget.cpp

```diff
@@ -0,0 +1,834 @@
+  if (controller_name_.empty())
+  {
+    QMessageBox::warning(this, "Error Saving", "A name must be given for the controller!");
+    return false;
+  }
+
+  // Check if this is an existing controller
+  if (!current_edit_controller_.empty())
+  {
+    // Find the controller we are editing based on the goup name string
+    searched_controller = config_data_->findControllerByName(current_edit_controller_);
+  }
+
+  // Check that the controller name is unique
+  for (std::vector<Controller>::const_iterator controller_it = config_data_->controllers_config_.begin();
+       controller_it != config_data_->controllers_config_.end(); ++controller_it)
+  {
+    if (controller_it->name_.compare(controller_name_) == 0)  // the names are the same
+    {
+      // is this our existing controller? check if controller pointers are same
+      if (&(*controller_it) != searched_controller)
+      {
+        QMessageBox::warning(this, "Error Saving", "A controller already exists with that name!");
+        return false;
+      }
+    }
+  }
+
+  adding_new_controller_ = false;
+
+  // Save the new controller name or create the new controller
+  if (searched_controller == NULL)  // create new
+  {
+    Controller new_controller;
+    new_controller.name_ = controller_name_;
+    new_controller.type_ = controller_type_;
+    config_data_->controllers_config_.push_back(new_controller);
+    adding_new_controller_ = true;  // remember this controller is new
+  }
+  else
+  {
+    // Remember old controller name and type
+    const std::string old_controller_name = searched_controller->name_;
+
+    // Change controller name
+    searched_controller->name_ = controller_name_;
+    searched_controller->type_ = controller_type_;
+  }
+
+  // Reload main screen table
+  loadControllersTree();
+
+  // Update the current edit controller so that we can proceed to the next screen, if user desires
+  current_edit_controller_ = controller_name_;
+
+  return true;
+}
+
+// ******************************************************************************************
+// Edit whatever element is selected in the tree view
+// ******************************************************************************************
+void ControllersWidget::editSelected()
+{
+  QTreeWidgetItem* item = controllers_tree_->currentItem();
+  QTreeWidgetItem* controller_item;
+  // Check that something was actually selected
+  if (item == NULL)
+    return;
+
+  adding_new_controller_ = false;
+
+  int type_ = item->data(0, Qt::UserRole).value<int>();
+
+  // Get the user custom properties of the currently selected row
+  if (type_ == 2)
+  {
+    // To which controller does this joint belong
+    controller_item = item->parent()->parent();
+    current_edit_controller_ = controller_item->text(0).toUtf8().constData();
+
+    Controller* this_controller = config_data_->findControllerByName(current_edit_controller_);
+    // Load the data
+    loadJointsScreen(this_controller);
```

Comment:
this is the same as ``type_ ==  1``. Does type 2 apply for this controllers screen? Should it be removed?

Reply:
The type two are the individual `joints`, so this means if what was selected before the edit was a joint, the user should be directed to `joints edit screen`.